### PR TITLE
feat: add GitHub Actions workflow YAML scanning

### PR DESCRIPTION
**Pull request by kotakanbe**
## Summary

- Add `--file .github/workflows/ci.yml` support to scan GitHub Actions for lifecycle health
- New `ghaworkflow` parser extracts `owner/repo` from `uses:` directives and evaluates them as GitHub repositories via existing analysis pipeline
- No changes to domain types or application layer — standalone infrastructure parser routes GitHub URLs through existing `RunFromPURLs`

### Design decision

Architect review concluded that the workflow parser should NOT implement `DependencyParser` (which is PURL-centric). Instead, it is a standalone function that outputs GitHub URLs, routed directly from the CLI layer to `RunFromPURLs(nil, githubURLs)`. This preserves `ParsedDependency` domain type purity.

### Supported `uses:` patterns

| Pattern | Example | Result |
|---|---|---|
| `owner/repo@ref` | `actions/checkout@v4` | `https://github.com/actions/checkout` |
| `owner/repo/path@ref` | `github/codeql-action/init@v4` | `https://github.com/github/codeql-action` |
| Reusable workflow | `owner/repo/.github/workflows/ci.yml@main` | `https://github.com/owner/repo` |
| `docker://...` | `docker://alpine:3.8` | Skipped |
| `./local` | `./.github/actions/my-action` | Skipped |

## Integration test result

Tested against this repository's own CI workflow (`.github/workflows/ci.yml`):

```
$ uzomuzo scan --file .github/workflows/ci.yml

VERDICT  PURL                                              LIFECYCLE      EOL
ok       https://github.com/actions/checkout               Legacy-Safe    Not EOL
review   https://github.com/actions/setup-go               Review Needed  Unknown
ok       https://github.com/golangci/golangci-lint-action  Active         Not EOL

Summary: 3 dependencies | 2 ok | 0 caution | 0 replace | 1 review
```

**Notes:**
- `actions/checkout` — resolved to npm package `checkout` (Legacy-Safe). The GitHub Action itself is actively maintained; the npm resolution is a known limitation where deps.dev maps the repo to the wrong package ecosystem.
- `actions/setup-go` — Review Needed because deps.dev cannot find a matching package for this Action.
- `golangci/golangci-lint-action` — Active, correctly resolved via Go module registry.

These results highlight that GitHub Actions do not have a native PURL ecosystem, so resolution depends on whatever package the repository also publishes (if any). This is an area for future improvement.

## Test plan

- [x] Unit tests for all `uses:` patterns (standard, monorepo, docker, local, reusable)
- [x] Unit tests for `IsWorkflowYAML` detection (path-based + content-based)
- [x] Edge cases: invalid YAML, whitespace, SHA-pinned actions, deep subpaths
- [x] `go test ./...` passes
- [x] `golangci-lint run` — 0 issues
- [x] Integration test: `uzomuzo scan --file .github/workflows/ci.yml` (see above)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/interfaces/cli/scan.go [hunk 1]
```diff
@@ -8,13 +8,15 @@ import (
...
+	"path/filepath"
 	"strings"
 	"time"
 
 	scanapp "github.com/future-architect/uzomuzo-oss/internal/application/scan"
 	domaincfg "github.com/future-architect/uzomuzo-oss/internal/domain/config"
 	"github.com/future-architect/uzomuzo-oss/internal/domain/depparser"
 	domainscan "github.com/future-architect/uzomuzo-oss/internal/domain/scan"
+	"github.com/future-architect/uzomuzo-oss/internal/infrastructure/depparser/ghaworkflow"
 )
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/interfaces/cli/scan.go]: Fixed in 099f9f3 — moved workflow detection/parsing to composition root (cmd/uzomuzo) via injected `WorkflowDetector` and `WorkflowParser` function types, removing the Infrastructure import from the Interfaces layer.

File: internal/interfaces/cli/scan.go [hunk 2]
```diff
@@ -268,6 +279,48 @@ func finalizeScanOutput(svc *scanapp.Service, result *scanapp.Result, opts ScanO
...
 }
 
+// runScanWorkflow parses a GitHub Actions workflow YAML and evaluates referenced Actions.
+func runScanWorkflow(ctx context.Context, svc *scanapp.Service, data []byte, opts ScanOptions, policy domainscan.FailPolicy) error {
+	githubURLs, err := ghaworkflow.ParseGitHubURLs(data)
+	if err != nil {
+		return fmt.Errorf("failed to parse workflow: %w", err)
+	}
+	if len(githubURLs) == 0 {
+		return fmt.Errorf("no GitHub Actions references found in workflow file")
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/interfaces/cli/scan.go]: Fixed in 099f9f3 — now returns empty result with a log message instead of an error, aligning with other structured parsers (go.mod/SBOM).

File: internal/infrastructure/depparser/ghaworkflow/parser.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+// hasWorkflowMarkers checks whether the byte prefix contains top-level YAML keys
+// that indicate a GitHub Actions workflow file: "on:" (or "true:" from YAML boolean
+// parsing of bare `on`) and "jobs:".
+func hasWorkflowMarkers(data []byte) bool {
+	s := string(data)
+	hasOn := strings.Contains(s, "\non:") || strings.HasPrefix(s, "on:") ||
+		strings.Contains(s, "\n\"on\":") || strings.HasPrefix(s, "\"on\":")
+	hasJobs := strings.Contains(s, "\njobs:") || strings.HasPrefix(s, "jobs:")
+	return hasOn && hasJobs
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/depparser/ghaworkflow/parser.go]: Fixed in 099f9f3 — updated comment to accurately describe the heuristic (checks for "on:" and quoted "on":, not "true:").

**Comment by kotakanbe** [comment]
### Copilot Review Resolution (`--copilot-only` mode) 

- [x] Read review prompt and gather context
- [x] Classify 4 Copilot review comments (Round 2)
- [ ] Fix classified issues
- [ ] Build / test / lint
- [ ] Commit and push
- [ ] Reply and resolve threads
- [ ] Detect recurring patterns and update rules
- [ ] Post final summary

---

### Classification (Round 2)

| # | File | Classification | Rationale |
|---|------|---------------|-----------|
| 1 | `commands.go:168` | **FIX** | `detectWorkflowFile` reads entire file for non-workflow YAML; optimize with prefix-only read |
| 2 | `scan.go:149` | **FIX** | `--sample`/`--line-range` silently ignored for workflows; must reject per existing rule |
| 3 | `parser.go:60` | **FIX** | Map iteration yields non-deterministic URL order; sort job keys |
| 4 | `scan.go:53` | **FIX** | Doc comment missing workflow YAML in `--file` description |

*Applying fixes...*

File: cmd/uzomuzo/commands.go [hunk 4]
```diff
@@ -138,7 +142,30 @@ func scanAction(ctx context.Context, cfg *domaincfg.Config, cmd *urfcli.Command)
...
+	}
+
+	prefix := data
+	if len(prefix) > 1024 {
+		prefix = prefix[:1024]
+	}
+	if ghaworkflow.IsWorkflowYAML(filePath, prefix) {
+		return data, true, nil
+	}
+	return nil, false, nil
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply cmd/uzomuzo/commands.go]: Fixed in 1b63d0d — optimized to read only a small prefix for non-workflow YAML files. Path-based detection is tried first, then a 1024-byte prefix read for marker detection. The full file is read only after confirming it is a workflow.

File: internal/interfaces/cli/scan.go [hunk 5]
```diff
@@ -108,7 +116,7 @@ func runScanSBOM(ctx context.Context, svc *scanapp.Service, opts ScanOptions, pa
 }
 
 // runScanFile handles --file input (go.mod, SBOM, or PURL/URL list).
-func runScanFile(ctx context.Context, svc *scanapp.Service, opts ScanOptions, parsers map[string]depparser.DependencyParser, policy domainscan.FailPolicy, detectFile FileDetector) error {
+func runScanFile(ctx context.Context, svc *scanapp.Service, opts ScanOptions, parsers map[string]depparser.DependencyParser, policy domainscan.FailPolicy, detectFile FileDetector, detectWorkflow WorkflowDetector, parseWorkflow WorkflowParser) error {
 	filePath := opts.Filename
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/interfaces/cli/scan.go]: Fixed in 036da88.

File: internal/infrastructure/depparser/ghaworkflow/parser_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			got := ghaworkflow.IsWorkflowYAML(tt.filePath, tt.prefix)
+			if got != tt.want {
+				t.Errorf("IsWorkflowYAML(%q) = %v, want %v", tt.filePath, got, tt.want)
+			}
+		})
+	}
+}
+
+// assertURLsEqual compares two URL slices as unordered sets (YAML map iteration is non-deterministic).
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/depparser/ghaworkflow/parser_test.go]: Fixed in 036da88 — updated comment to reflect contract-based set comparison.
